### ISSUE TITLE
false positive unnecessary parentheses in conjunction with lambdas

### ISSUE THREAD

**Issue by robstoll** [arturbosch/detekt#1222]
## Expected Behavior
I would argue that one does not put parentheses around lambdas if it is not necessary. Do not report it if lambdas are involved because in most cases it is necessary due to ambiguity problems.

## Current Behavior
A lot of issues (depending on the code of course) which cannot be solved and need to be ignored.

## Steps to Reproduce (for bugs)
see project link

## Your Environment
* Version used: 1.0.0.RC8
* Link to your project: https://github.com/loewenfels/dep-graph-releaser

# Report
I am ignoring the rule thus no link to sonarcloud but at least a list for easier reproduction. Only the first and the last issue can be resolved, all others have to stay the way they are
```
C:\projects\dep-graph-releaser\api\dep-graph-releaser-api-common\src\main\kotlin\ch\loewenfels\depgraph\data\ReleasePlan.kt:102:60 
Parentheses in (it.isSubmodule && it.level == project.level) are unnecessary and can be replaced with: it.isSubmodule && it.level == project.level
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:155:57 
Parentheses around the lambda ({ isA {} }) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:175:57 
Parentheses around the lambda ({ isA { isStateReady() property(subject::nextDevVersion).toBe(rootProjectIdAndVersions.nextDevVersion) } }) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:288:57 
Parentheses around the lambda ( { isJenkinsMavenReleaseWaiting(project.nextDevVersion, dependency) } ) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:313:57 
Parentheses around the lambda ( { isJenkinsUpdateDependencyWaiting(dependency) } ) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:346:57 
Parentheses around the lambda ( { isJenkinsMavenReleaseWaiting(project.nextDevVersion, syntheticRoot) } ) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:428:57 
Parentheses around the lambda ( { isJenkinsMultiMavenReleaseWaiting(project.nextDevVersion, dependency) } ) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:534:54 
Parentheses around the lambda ({ contains("cyclic dependencies", dependencyBranch, *otherDependencyBranches) }) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\atrium\src\main\kotlin\ch\tutteli\atrium\spekProjectAssertions.kt:546:51 
Parentheses around the lambda ({ contains("cyclic dependencies", dependencyBranch, *otherDependencyBranches) }) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\dep-graph-releaser-maven\src\test\kotlin\ch\loewenfels\depgraph\maven\IntegrationSpec.kt:174:62 
Parentheses around the lambda ({ contains(pom.absolutePathAsString) }) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\dep-graph-releaser-maven\src\test\kotlin\ch\loewenfels\depgraph\maven\IntegrationSpec.kt:184:62 
Parentheses around the lambda ({ contains(pom.absolutePathAsString) }) are unnecessary and can be removed.
C:\projects\dep-graph-releaser\dep-graph-releaser-runner\src\main\kotlin\ch\loewenfels\depgraph\runner\Orchestrator.kt:76:22 
Parentheses in (length - this.length) are unnecessary and can be replaced with: length - this.length`
```

**Comment by Mauin** [comment]
Thanks for reporting the issue. Can you post some simple sample code that reproduces this issue? Just looking at the logs I'd like to have a bit more context.

**Comment by robstoll** [comment]
Use the link of the repo in combination with the log there are all simple examples:
https://github.com/loewenfels/dep-graph-releaser/blob/master/atrium/src/main/kotlin/ch/tutteli/atrium/spekProjectAssertions.kt#L155

**Comment by marschwar** [comment]
In `.filter { it.level == project.level + 1 || (it.isSubmodule && it.level == project.level) }` and `(1..(length - this.length) step padString.length).forEach { _ ->` the parentheses are really not needed

There is a false positive with vararg lambdas though:
```
it("should not report if parentheses are needed") {
	val code = """
		fun foo(vararg other: () -> Unit): Unit = TODO()

		fun bar() {
			// this will report a violation
			foo ({  })

			// this will not compile
			// foo {  }
		}
	""".trimIndent()
	assertThat(subject.lint(code)).hasSize(0)
}
```

**Comment by Mauin** [comment]
Unnecessary parentheses are definitely a rule that can come in handy sometimes, but sometimes you really want to add readability to the expression you're writing. I'm not quite sure how to proceed. In general we aim to have the `UnnecessaryParentheses` rule report in the same way that the IntelliJ inspection does. 
@robstoll Does IntelliJ flag these parentheses in the same way?

**Comment by robstoll** [comment]
No, IntelliJ does not flag them. So far I have never seen IntelliJ flagging some of my code. Using parentheses to enhance readability or to make things crystal clear (not everyone knows operator precedence by heart). So I guess it only flags things like:
```
val a = (1 + 1)
val b = 1 + ((1 +1))
```

**Comment by schalkms** [comment]
I'm actually for removing or deprecating this rule.
This rule has raised so many controversies. In the end, I think this rule doesn't bring much to the table.
I could also live we a new rule set called `controversial`. Other tools also have that.
`UnnecessaryParentheses` would be a good candidate for this rule set.

My preference would be to remove this rule altogether.

CC @arturbosch

**Comment by arturbosch** [comment]
I wouldn't remove the rule as it has some usages for simple cases. However a `controversial` rule set sounds good to me.
We could also make the rule ignore cases when direct lambdas are involved?

**Comment by schalkms** [comment]
> We could also make the rule ignore cases when direct lambdas are involved?

Good idea. In any other case I wouldn't spend more time in implementing other config options (like mentioned in issue #1592).
